### ISSUE TITLE
`randint(max)` causes a graph break, but not `rand().mul(max).floor().to(torch.long)` (on CPU)

### ISSUE THREAD

**Issue by vmoens** [pytorch/pytorch#135664]
### 🐛 Describe the bug

The following code causes a graph break (on CPU):
```python
    def policy(obs, epsilon):
        q_values = q_network_detach(obs)
        actions = torch.argmax(q_values, dim=1)
        actions_random = torch.rand(actions.shape, device=actions.device).mul(n_act).floor().to(torch.long)
        # actions_random = torch.randint_like(actions, n_act)
        use_policy = torch.rand(actions.shape, device=actions.device).gt(epsilon)
        return torch.where(use_policy, actions, actions_random)
```
but not the same function with the first version of `action_random` uncommented (which supposedly does the same job)

The graph break reads:
```
W0911 10:15:45.971000 28832 torch/_dynamo/exc.py:284] [0/0] Backend compiler failed with a fake tensor exception at 
W0911 10:15:45.971000 28832 torch/_dynamo/exc.py:284] [0/0]   File "/Users/vmoens/Repos/rl/leanrl/leanrl/dqn_torchcompile.py", line 208, in policy
W0911 10:15:45.971000 28832 torch/_dynamo/exc.py:284] [0/0]     return torch.where(use_policy, actions, actions_random)
W0911 10:15:45.971000 28832 torch/_dynamo/exc.py:284] [0/0] Adding a graph break.
[...]
  File "/Users/vmoens/venv/rl/lib/python3.11/site-packages/torch/_subclasses/fake_impls.py", line 147, in dispatch_to_op_implementations_dict
    return op_implementations_dict[func](fake_mode, func, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/vmoens/venv/rl/lib/python3.11/site-packages/torch/_subclasses/fake_impls.py", line 386, in local_scalar_dense
    raise DataDependentOutputException(func)
torch._subclasses.fake_tensor.DataDependentOutputException: aten._local_scalar_dense.default
[...]
torch._dynamo.exc.Unsupported: Backend compiler failed with a fake tensor exception at 
  File "/Users/vmoens/Repos/rl/leanrl/leanrl/dqn_torchcompile.py", line 208, in policy
    return torch.where(use_policy, actions, actions_random)
Adding a graph break.
```

### Versions

nightly

cc @chauhang @penguinwu @eellison @ezyang @bobrenjc93 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @amjames @zou3519 @bdhirsh @rec

**Comment by masnesral** [comment]
Looks like all the labels already applied. Assuming this is not hi-pri and doesn't need discussion at pt2 triage review.

**Comment by StrongerXi** [comment]
Got a repro on main:
```python
import torch

@torch.compile(backend="aot_eager", fullgraph=True)
def f(actions, n_act, epsilon=0.1):
    # Works
    #actions_random = torch.rand(actions.shape, device=actions.device).mul(n_act).floor().to(torch.long)

    # Breaks
    actions_random = torch.randint_like(actions, n_act)

    use_policy = torch.rand(actions.shape, device=actions.device).gt(epsilon)
    return torch.where(use_policy, actions, actions_random)

x = torch.ones([1], dtype=torch.int64)
y = torch.tensor(5)
f(x, y)
```

**Comment by StrongerXi** [comment]
Okay, `FakeTensor.item_memo` got lost after fake mode epoch bump:
https://github.com/pytorch/pytorch/blob/a952f42bdb71585d2308fde71884778cc509cc53/torch/_functorch/_aot_autograd/traced_function_transforms.py#L901-L904

Removing the `SymFloat` check here fixed it for me:
https://github.com/pytorch/pytorch/blob/a952f42bdb71585d2308fde71884778cc509cc53/torch/_subclasses/fake_tensor.py#L604-L606

But that check was added in #139564, which looks like some hot fix.

**Comment by bobrenjc93** [comment]
So the problem here is `randint_like` actually takes in an int argument for `high` but in this code we are passing a tensor. We have a clever boxed implementation that takes in a tensor `high` and calls item on it to get the integer. The problem though is that now this is a data dependent operation and we necessarily have to graph break.

The easiest way to fix this is to simply pass in an int instead of a tensor, eg:

```
import torch

@torch.compile(backend="aot_eager", fullgraph=True)
def f(actions, n_act, epsilon=0.1):
    # Works
    #actions_random = torch.rand(actions.shape, device=actions.device).mul(n_act).floor().to(torch.long)

    # Breaks
    actions_random = torch.randint_like(actions, n_act)

    return actions_random

x = torch.ones([1], dtype=torch.int64)
y = 5
f(x, y)
```

That being said, it does feel like we should be able to handle this case more gracefully. The C++ codegen/dispatcher/boxed stuff is still quite foreign to me so perhaps @bdhirsh can chime in with thoughts. I think this compile problem generalizes to any operator that takes in an int but is given a scalar tensor instead.

**Comment by zou3519** [comment]
Hot take: all of our operators that take int values (to be distinguished from int dim) should also take in a tensor and the tensor variant should not call .item before the backend implementations. Maybe there is some cool way of easily getting this up without sitting down and rewriting all of those operators

**Comment by bdhirsh** [comment]
from pt2-dispatcher triage:

(1) create a new overload for `randint_like` that takes in max as a tensor. probably need to deal with all existing overloads of this op

(2) profit

optional things to think through later:

(1) have the impl for this tensor overload not do a h2d sync to make cudagraphs happy

(2) audit other ops that have this behavior. if there are enough of them, we could do something tag-based that is more automatic
